src/graph/FetchEdgesExecutor.cpp

```diff
@@ -0,0 +1,340 @@
+        auto aliasProps = expCtx_->aliasProps();
+        for (auto pair : aliasProps) {
+            if (pair.first != *label) {
+                status = Status::SyntaxError(
+                    "[%s.%s] not declared in %s.",
+                    pair.first.c_str(), pair.second.c_str(), (*label).c_str());
+                break;
+            }
+        }
+    } while (false);
+
+    return status;
+}
+
+Status FetchEdgesExecutor::setupColumns() {
+    auto *label = sentence_->label();
+    auto schema = ectx()->schemaManager()->getEdgeSchema(spaceId_, edgeType_);
+    auto iter = schema->begin();
+    if (yieldColsHolder_ == nullptr) {
+        yieldColsHolder_ = std::make_unique<YieldColumns>();
+    }
+    while (iter) {
+        auto *name = iter->getName();
+        auto *ref = new std::string("");
+        auto *labelName = new std::string(*label);
+        Expression *expr = new AliasPropertyExpression(ref, labelName, new std::string(name));
+        YieldColumn *column = new YieldColumn(expr);
+        yieldColsHolder_->addColumn(column);
+        yields_.emplace_back(column);
+        ++iter;
+    }
+    return Status::OK();
+}
+
+void FetchEdgesExecutor::execute() {
+    FLOG_INFO("Executing FetchEdges: %s", sentence_->toString().c_str());
+    auto status = setupEdgeKeys();
+    if (!status.ok()) {
+        DCHECK(onError_);
+        onError_(std::move(status));
+        return;
+    }
+
+    if (edgeKeys_.empty()) {
+        onEmptyInputs();
+    }
+
+    fetchEdges();
+}
+
+Status FetchEdgesExecutor::setupEdgeKeys() {
+    Status status = Status::OK();
+    if (sentence_->isRef()) {
+        status = setupEdgeKeysFromRef();
+    } else {
+        status = setupEdgeKeysFromExpr();
+    }
+
+    return status;
+}
+
+Status FetchEdgesExecutor::setupEdgeKeysFromRef() {
+    const InterimResult *inputs;
+    if (sentence_->ref()->isInputExpr()) {
+        inputs = inputs_.get();
+        if (inputs == nullptr) {
+            // we have empty imputs from pipe.
+            return Status::OK();
+        }
+    } else {
+        inputs = ectx()->variableHolder()->get(varname_);
+        if (inputs == nullptr) {
+            return Status::Error("Variable `%s' not defined", varname_.c_str());
+        }
+    }
+
+
+    auto ret = inputs->getVIDs(*srcid_);
+    if (!ret.ok()) {
+        return ret.status();
+    }
+    auto srcVids = std::move(ret).value();
+
+    ret = inputs->getVIDs(*dstid_);
+    if (!ret.ok()) {
+        return ret.status();
+    }
+    auto dstVids = std::move(ret).value();
+
+    std::vector<int64_t> ranks;
+    if (rank_ != nullptr) {
+        ret = inputs->getVIDs(*rank_);
+        if (!ret.ok()) {
+            return ret.status();
+        }
+        ranks = std::move(ret).value();
+    }
+
+    for (auto index = 0u; index < srcVids.size(); ++index) {
+        storage::cpp2::EdgeKey key;
+        key.set_src(srcVids[index]);
+        key.set_edge_type(edgeType_);
+        key.set_dst(dstVids[index]);
```

Comment:
Is dstVids.size() always same with srcVids.size()?

Reply:
Yes, I made some restrictions. They would come from same variable or input.